# Chapter 12. Deep Computer Vision Using Convolutional Neural Networks

CNNs were inspired by the visual cortex, where neurons respond to patterns within small local receptive fields. Higher-level neurons combine lower-level patterns to detect increasingly complex features.

## Convolutional Layers

Convolutional layers connect each neuron only to a local receptive field instead of every input pixel.

This local connectivity allows early layers to detect simple features, such as edges and lines, while deeper layers combine them into larger and more complex patterns.

Unlike fully connected networks, CNNs preserve the spatial structure of images instead of flattening them immediately.

This makes CNNs much more efficient for larger images, since fully connected layers would require an extremely large number of parameters.

Zero padding adds zeros around the input to preserve more of its spatial dimensions.

The `stride` controls how far the receptive field moves between positions, and a larger stride produces smaller output feature maps with lower computational cost.

### Filters

A filter, also called a kernel, is a small set of weights that slides across the input.

Each filter learns to detect a particular pattern, such as an edge, line, or texture.

Applying one filter across the whole input produces a feature map.

The feature map shows where that pattern is detected most strongly.

### Stacking Multiple Feature Maps

A convolutional layer usually contains multiple filters.

Each filter produces one feature map, so the layer output is a stack of multiple feature maps.

All neurons within the same feature map share the same filter weights and bias.

This weight sharing greatly reduces the number of parameters and allows the same feature to be detected at different locations.

For one input instance, the output shape of a convolutional layer is:

`[out_channels, output_height, output_width]`

Each output channel is one feature map, and `output_height` and `output_width` describe the spatial dimensions of that feature map.

With a batch dimension, PyTorch uses:

`[batch_size, out_channels, output_height, output_width]`

### Implementing Convolutional Layers with PyTorch

First, we load two sample RGB images, stack them into one NumPy array, convert them to a `float32` tensor, and scale the pixel values from `0–255` to `0–1`.

In [12]:
import numpy as np
import torch
from sklearn.datasets import load_sample_images

sample_images = np.stack(load_sample_images()["images"])
sample_images = torch.tensor(sample_images, dtype=torch.float32) / 255

In [13]:
sample_images.shape

torch.Size([2, 427, 640, 3])

The tensor has shape:

`[batch_size, height, width, channels]`

PyTorch convolutional layers expect image tensors in the format:

`[batch_size, channels, height, width]`

so the dimensions must be rearranged using `permute()`.

In [15]:
sample_images_permuted = sample_images.permute(0, 3, 1, 2)
sample_images_permuted.shape

torch.Size([2, 3, 427, 640])

Next, the images are center-cropped to a smaller spatial size.

In [17]:
import torchvision
import torchvision.transforms.v2 as T

cropped_images = T.CenterCrop((70, 120))(sample_images_permuted)
cropped_images.shape

torch.Size([2, 3, 70, 120])

PyTorch provides `nn.Conv2d` for 2D convolutional layers.

Here, `out_channels=32` means the layer has `32` filters, and `kernel_size=7` means that each filter has spatial dimensions `7 × 7`.

The term “2D” refers to the two spatial dimensions, height and width.

However, `nn.Conv2d` takes a 4D tensor because the batch and channel dimensions are also included.

In [19]:
import torch.nn as nn

torch.manual_seed(42)

conv_layer = nn.Conv2d(in_channels=3, out_channels=32, kernel_size=7)
fmaps = conv_layer(cropped_images)

In [20]:
fmaps.shape

torch.Size([2, 32, 64, 114])

The output shape is:

`[2, 32, 64, 114]`

The number of channels increases from `3` to `32` because each of the `32` filters produces one output feature map.

The spatial dimensions shrink because the default setting is `padding=0`. With a `7 × 7` kernel and `stride=1`, the height and width each decrease by `6`.

In [22]:
conv_layer = nn.Conv2d(in_channels=3, out_channels=32, kernel_size=7,
                       padding="same")
fmaps = conv_layer(cropped_images)
fmaps.shape

torch.Size([2, 32, 70, 120])

With `padding="same"`, the output spatial dimensions remain `70 × 120`.

A larger `stride` reduces the output size. For example, with `stride=2`, `padding=3`, and `kernel_size=7`, the spatial dimensions are approximately halved.

The filters and biases of an `nn.Conv2d` layer are stored in the `weight` and `bias` attributes.

In [25]:
conv_layer.weight.shape

torch.Size([32, 3, 7, 7])

In [26]:
conv_layer.bias.shape

torch.Size([32])

The weight tensor has shape:

`[out_channels, in_channels, kernel_height, kernel_width]`

The bias tensor has shape:

`[out_channels]`

The input image height and width do not appear in the filter shape because the same filter weights are shared across all spatial locations.

A convolutional layer performs a linear operation, so an activation function should normally be added after it.

Without activation functions, stacking several convolutional layers would still be equivalent to one linear convolutional operation.

The weights may also be reinitialized to match the activation function, such as using He initialization with `ReLU`.

## Pooling Layers

Pooling layers shrink feature maps to reduce computation, memory usage, and the risk of overfitting.

Unlike convolutional layers, pooling layers have no weights or biases. They simply aggregate values within each receptive field using an operation such as the maximum or mean.

Max pooling is the most common type of pooling.

For example, a `2 × 2` pooling kernel with `stride=2` keeps only the maximum value from each receptive field, reducing the height and width by about half.

Pooling is usually applied independently to each input channel, so the number of channels remains unchanged.

Max pooling provides some invariance to small translations, which can be useful for classification.

However, it is destructive because it discards many input values. For tasks such as semantic segmentation, equivariance is often more important: when the input shifts, the output should shift correspondingly.

## Implementing Pooling Layers with PyTorch

`nn.MaxPool2d` keeps the maximum value within each spatial region, while `nn.AvgPool2d` keeps the average.

Max pooling is more commonly used because it preserves the strongest features and provides greater translation invariance.

In [50]:
max_pool = nn.MaxPool2d(kernel_size=2) # 2x2 kernel

The stride defaults to the kernel size, and the padding is set to `0` by default.

Pooling can also be applied across the channel dimension instead of the spatial dimensions.

Depthwise max pooling keeps the strongest activation among groups of feature maps, which can help the model become invariant to feature variations such as rotation, thickness, or brightness.

In [69]:
import torch.nn.functional as F

class DepthPool(torch.nn.Module):
    def __init__(self, kernel_size, stride=None, padding=0):
        super().__init__()
        self.kernel_size = kernel_size
        self.stride = stride if stride is not None else kernel_size
        self.padding = padding

    def forward(self, inputs):
        batch, channels, height, width = inputs.shape
        Z = inputs.view(batch, channels, height * width) # merge spatial dims
        Z = Z.permute(0, 2, 1) # switch spatial and channels dim
        Z = F.max_pool1d(Z, kernel_size=self.kernel_size, stride=self.stride,
            padding=self.padding)
        Z = Z.permute(0, 2, 1)
        return Z.view(batch, -1, height, width)

Global average pooling computes one average value for each entire feature map.

It removes the remaining spatial information and is often used near the output of a CNN.

In [62]:
global_avg_pool = nn.AdaptiveAvgPool2d(output_size=1)
output = global_avg_pool(cropped_images)

Alternatively, you could use the `torch.mean()` to get the same output.

In [64]:
output = cropped_images.mean(dim=(2, 3), keepdim=True)

## CNN Architectures

A typical CNN stacks convolutional layers followed by activation functions and pooling layers.

As the network gets deeper, the spatial dimensions usually become smaller, while the number of feature maps increases.

Near the output, the feature maps are flattened and passed to fully connected layers for the final prediction.

Stacking two convolutional layers with `3 × 3` kernels is often preferred over using one `5 × 5` kernel because it usually requires fewer parameters and computations.

In [84]:
from functools import partial

device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

DefaultConv2d = partial(nn.Conv2d, kernel_size=3, padding="same")

model = nn.Sequential(
    DefaultConv2d(in_channels=1, out_channels=64, kernel_size=7), nn.ReLU(),
    nn.MaxPool2d(kernel_size=2),
    DefaultConv2d(in_channels=64, out_channels=128), nn.ReLU(),
    DefaultConv2d(in_channels=128, out_channels=128), nn.ReLU(),
    nn.MaxPool2d(kernel_size=2),
    DefaultConv2d(in_channels=128, out_channels=256), nn.ReLU(),
    DefaultConv2d(in_channels=256, out_channels=256), nn.ReLU(),
    nn.MaxPool2d(kernel_size=2),
    nn.Flatten(),
    nn.Linear(in_features=2304, out_features=128), nn.ReLU(),
    nn.Dropout(0.5),
    nn.Linear(in_features=128, out_features=64), nn.ReLU(),
    nn.Dropout(0.5),
    nn.Linear(in_features=64, out_features=10),
).to(device)

`DefaultConv2d` uses `kernel_size=3` and `padding="same"` as default arguments, reducing repeated code.

The first layer uses a larger `7 × 7` kernel, while later layers use smaller `3 × 3` kernels.

Pooling layers reduce the spatial dimensions:

`28 × 28 → 14 × 14 → 7 × 7 → 3 × 3`

At the same time, the number of feature maps increases:

`64 → 128 → 256`

Before the first fully connected layer, the tensor contains `256` feature maps of size `3 × 3`.

After flattening, this gives:

$$
256 \times 3 \times 3 = 2304
$$

input features.

The output layer has `10` units because Fashion MNIST has `10` classes.

The model outputs logits without a softmax layer, so it should be trained with `nn.CrossEntropyLoss()`.

`nn.LazyLinear` can be used instead of manually calculating `in_features`.

It automatically determines the correct number of input features the first time it receives data.

### LeNet-5

LeNet-5 is an early CNN architecture developed for handwritten digit recognition.

It follows the now-familiar pattern of convolutional layers and pooling layers followed by fully connected layers, but it uses `tanh`, average pooling, and an RBF output instead of the `ReLU`, max pooling, and softmax-style outputs commonly used in modern CNNs.

### AlexNet

AlexNet was a much larger and deeper CNN than LeNet-5 and achieved a major improvement in the 2012 ImageNet competition.

Unlike earlier CNNs, it stacked multiple convolutional layers before applying pooling, allowing the network to learn more complex hierarchical features.

AlexNet introduced several practices that became influential in large-scale image classification:

- Stacking multiple convolutional layers before pooling
- Using dropout in fully connected layers
- Applying data augmentation to reduce overfitting
- Training a large CNN efficiently on GPUs

#### Data Augmentation

Data augmentation generates realistic variations of existing training images to increase the effective size and diversity of the training set.

Common transformations include shifting, cropping, rotating, resizing, horizontal flipping, and adjusting color or lighting conditions.

Because the model sees different versions of the same object during training, it becomes less sensitive to changes in position, orientation, size, and lighting.

The transformations should preserve the original class and produce realistic images rather than arbitrary noise.

Data augmentation is generally applied only during training.

Test-time augmentation applies several transformations to a test image and combines their predictions, although this requires additional computation.

### GoogLeNet

GoogLeNet introduced inception modules, which allowed the network to become much deeper while using parameters more efficiently than earlier CNNs.

An inception module processes the same input through several parallel branches and concatenates their outputs along the channel dimension.

The parallel branches use different operations, such as:

- `1 × 1` convolution
- `3 × 3` convolution
- `5 × 5` convolution
- Max pooling

Because all branches preserve the same height and width, their output feature maps can be concatenated using `torch.cat(..., dim=1)`.

Using different kernel sizes allows the module to capture patterns at multiple spatial scales.

The `1 × 1` convolutions also learn relationships across channels and act as bottleneck layers that reduce the number of feature maps before more expensive convolutions.

By reducing the channel dimension before applying larger kernels, bottleneck layers lower the number of parameters and computational cost.

Each sequence such as `1 × 1 → 3 × 3` can also learn more complex patterns than a single convolutional layer.

GoogLeNet stacks nine inception modules, with pooling layers used between some modules to reduce the spatial dimensions.

Near the output, global average pooling replaces several large fully connected layers, reducing the number of parameters and the risk of overfitting.

### ResNet

ResNet introduced skip connections, which add the input of a block directly to its output.

Instead of learning the full target mapping $h(x)$, the block learns a residual function $f(x)=h(x)-x$, so the final output becomes:

$$
h(x)=f(x)+x
$$

Skip connections allow signals and gradients to travel more easily through very deep networks.

At the beginning of training, when the residual branch produces values close to zero, the block behaves approximately like an identity function by passing its input forward.

A residual unit typically contains convolutional layers with batch normalization and `ReLU`, while the shortcut path bypasses these layers.

The residual branch output and shortcut input are added element-wise, so they must have the same shape.

When the spatial dimensions or number of channels change, the shortcut path uses a `1 × 1` convolution, often with `stride=2`, to match the output shape before addition.

### Xception

Xception replaces inception modules with depthwise separable convolutions.

A regular convolution learns spatial patterns and cross-channel patterns at the same time, while a depthwise separable convolution separates these two tasks.

A depthwise separable convolution has two steps:

1. **Depthwise convolution:** applies one spatial filter independently to each input channel.
2. **Pointwise convolution:** uses `1 × 1` convolutions to combine information across channels.

This separation usually requires fewer parameters, less memory, and fewer computations than a regular convolution.

However, depthwise separable convolutions are less suitable immediately after layers with very few channels, so Xception begins with regular convolutional layers.

In [108]:
class SeparableConv2d(nn.Module):
    def __init__(self, in_channels, out_channels, kernel_size, stride=1,
                 padding=0):
        super().__init__()
        self.depthwise_conv = nn.Conv2d(
            in_channels, in_channels, kernel_size, stride=stride,
            padding=padding, groups=in_channels)
        self.pointwise_conv = nn.Conv2d(
            in_channels, out_channels, kernel_size=1, stride=1, padding=0)

    def forward(self, inputs):
        return self.pointwiseconv(self.depthwise_conv(inputs))

Setting `groups=in_channels` creates one independent spatial filter per input channel.

The following `1 × 1` convolution then mixes these channel-specific features and produces the desired number of output channels.

### SENet

SENet adds a squeeze-and-excitation block to existing architectures such as Inception networks and ResNets.

An SE block learns which feature maps are important and recalibrates them by reducing irrelevant channels while preserving more useful ones.

An SE block has three main steps:

1. Global average pooling summarizes each feature map with one value.
2. A small fully connected network learns relationships among the channels.
3. A sigmoid output produces one importance score between `0` and `1` for each feature map.

Each feature map is multiplied by its corresponding score.

Feature maps with low scores are suppressed, while feature maps with scores close to `1` are largely preserved.

### Other Noteworthy Architectures

- **DenseNet:** Each layer receives the concatenated outputs of all earlier layers within the same dense block.
- **MobileNet:** Uses depthwise separable convolutions to reduce parameters and computation for mobile applications.
- **EfficientNet:** Jointly scales network depth, width, and input resolution using compound scaling.